src/store/undoRedo/index.ts

```diff
@@ -0,0 +1,143 @@
+import {ActionTree, GetterTree, Module, MutationTree} from 'vuex'
+import {IRootState} from '@/store'
+import {IFlowsState} from '@/store/flow'
+import {IBuilderState} from '@/store/builder'
+
+export interface IUndoRedoState {
+  snapshots: Snapshot[],
+  position: number,
+}
+
+export interface Snapshot {
+  flow: IFlowsState,
+  builder: IBuilderState,
+}
+
+export const stateFactory = (): IUndoRedoState => ({
+  snapshots: [],
+  position: -1,
+})
+
+export const getters: GetterTree<IUndoRedoState, IRootState> = {
+  canUndo(state): boolean {
+    const hasMoreThanInitialSnapshot = state.snapshots.length > 1
+    const isNotFirstPosition = state.position > 0
+    return hasMoreThanInitialSnapshot && isNotFirstPosition
+  },
+  canRedo(state): boolean {
+    const hasMoreThanInitialSnapshot = state.snapshots.length > 1
+    const isNotLastPosition = state.position < state.snapshots.length - 1
+    return hasMoreThanInitialSnapshot && isNotLastPosition
+  },
+}
+
+export const mutations: MutationTree<IUndoRedoState> = {
+  clearRedos(state) {
+    // clear everything after current position
+    state.snapshots.splice(state.position + 1)
+  },
+  addSnapshot(state, stateSnapshot: Snapshot) {
+    state.snapshots.push(stateSnapshot)
+    state.position += 1
+  },
+  undo(state) {
+    if (state.position > 0) {
+      state.position -= 1
+    } else {
+      console.warn('Cannot undo, we have already reached the beginning of the history')
+    }
+  },
+  redo(state) {
+    if (state.position < state.snapshots.length - 1) {
+      state.position += 1
+    } else {
+      console.warn('Cannot redo, we have already reached the end of the history')
+    }
+  },
```

Comment:
I kept the 'undo'/'redo' naming for these mutations, but to be more accurate, we might consider something like 'decrementPosition'/'incrementPosition' or 'decPosition'/'incPosition', although I don't like that very much :)